### ISSUE TITLE
False Positive Union Protocol Compatibility 

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5673]
**Describe the bug**
```python
from typing import TypeVar, Union

from _typeshed import SupportsLenAndGetItem

T = TypeVar("T")


def foo(x: SupportsLenAndGetItem[T]) -> T:
    ...


def bar(x: Union[list[int], list[str]]) -> None:
    foo(x) # False positive error
```

> error: Argument of type "list[int] | list[str]" cannot be assigned to parameter "x" of type "SupportsLenAndGetItem[T@foo]" in function "foo"
>     Type "list[int] | list[str]" cannot be assigned to type "SupportsLenAndGetItem[int]"
>       "list[str]" is incompatible with protocol "SupportsLenAndGetItem[int]"
>         "__getitem__" is an incompatible type
>           No overloaded function matches type "(__k: int, /) -> _T_co@SupportsLenAndGetItem" (reportGeneralTypeIssues)
> 

**VS Code extension or command-line**
pyright 1.1.321 introduced the error. 1.1.320 type checks successfully.

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. The problem was introduced when I fixed [this issue](https://github.com/microsoft/pyright/issues/5636). A fix will be included in the next release. I've added a unit test based on your code sample to prevent similar regressions in the future.

**Comment by erictraut** [comment]
This regression has affected several pyright users in various ways, so I decided to do a hot-fix release. I just published 1.1.322, which addresses the issue.
